test/integration/editor.spec.ts

```diff
@@ -0,0 +1,154 @@
+  expect(canvasPageOverlayBoundingBox).toBeDefined();
+
+  await canvasPageOverlayLocator.click({
+    position: {
+      x:
+        componentBoundingBox!.x + componentBoundingBox!.width / 2 - canvasPageOverlayBoundingBox!.x,
+      y:
+        componentBoundingBox!.y +
+        componentBoundingBox!.height / 2 -
+        canvasPageOverlayBoundingBox!.y,
+    },
+  });
+}
+
+test('can place new components from catalog', async ({ page }) => {
+  const appId = generateId();
+
+  await page.goto('/');
+  await createApp(page, `App ${appId}`);
+
+  const canvasFrameLocator = page.frameLocator(canvasFrame);
+
+  const componentCatalogLocator = page.locator(componentCatalog);
+  const canvasPageRootLocator = canvasFrameLocator.locator(pageRoot);
+  const canvasPageOverlayLocator = canvasFrameLocator.locator(pageOverlay);
+  const canvasInputLocator = canvasFrameLocator.locator('input');
+
+  await canvasPageRootLocator.waitFor();
+
+  await expect(canvasInputLocator).toHaveCount(0);
+
+  await componentCatalogLocator.hover();
+  await componentCatalogLocator
+    .locator(':has-text("TextField")[draggable]')
+    .dragTo(canvasPageOverlayLocator);
+
+  await expect(canvasInputLocator).toHaveCount(1);
+  await expect(canvasInputLocator).toBeVisible();
+
+  // Drag in a second component
+
+  await componentCatalogLocator.hover();
+  await componentCatalogLocator
+    .locator(':has-text("TextField")[draggable]')
+    .dragTo(canvasPageOverlayLocator);
+
+  await expect(canvasInputLocator).toHaveCount(2);
+});
+
+test('can move elements in page', async ({ page }) => {
+  const appId = generateId();
+
+  await page.goto('/');
+  await createApp(page, `App ${appId}`, JSON.stringify(twoFieldsDomTemplate));
+
+  const canvasFrameLocator = page.frameLocator(canvasFrame);
+
+  const canvasPageRootLocator = canvasFrameLocator.locator(pageRoot);
+  const canvasPageOverlayLocator = canvasFrameLocator.locator(pageOverlay);
+  const canvasInputLocator = canvasFrameLocator.locator('input');
+  const canvasMoveElementHandleLocator = canvasFrameLocator.locator(
+    ':has-text("TextField")[draggable]',
+  );
+
+  await canvasPageRootLocator.waitFor();
+
+  await canvasInputLocator.first().type('textField1');
+  await canvasInputLocator.nth(1).type('textField2');
+
+  await expect(canvasInputLocator.first()).toHaveAttribute('value', 'textField1');
+
+  // Move element by dragging
+
+  await expect(canvasMoveElementHandleLocator).not.toBeVisible();
+
+  const firstInputLocator = canvasInputLocator.first();
+  await selectComponent(page, firstInputLocator);
+
+  const canvasPageOverlayBoundingBox = await canvasPageOverlayLocator.boundingBox();
+
+  const secondTextFieldLocator = canvasInputLocator.nth(1);
+  const secondTextFieldBoundingBox = await secondTextFieldLocator.boundingBox();
+
+  expect(canvasPageOverlayBoundingBox).toBeDefined();
+  expect(secondTextFieldBoundingBox).toBeDefined();
+
+  await canvasMoveElementHandleLocator.dragTo(canvasPageOverlayLocator, {
```

Comment:
Can we use mouse events to simulate drag and drop instead? Did you try the strategy from https://github.com/mui/mui-toolpad/issues/758#issuecomment-1207914180 ?

Reply:
Probably can do done like those, I just thought `dragTo` would be better since it encapsulates the whole drag interaction?
Also I can add a comment here explaining that it's dragging to the right side of the other element.

Reply:
It seems to me that the method as used in the core is much closer to real user interaction, and less dependant on implementation details

Reply:
ok, ill also add a method about getting an (x,y) position relative to the overlay, it should make this kind of code more readable.

Reply:
Best would be to avoid the overlay whenever we can and rather detect/calculate based these things on the underlying app running in the iframe. The overlay is an implementation detail, we may want to change its implementation in the future. In fact I'd say it's very likely we will heavily change this implementation at some point.

Reply:
Yeah, I agree, the problem is that the overlay is handling all the events, I tried with other elements in the same position but it did not work. I'd be glad to do it in an alternative way but would have to look more into what else is possible.

Reply:
alright i hadn't realized we could use the Playwright's page object and do everything relative to it - that means i don't have to use the overlay after all!
thanks for the feedback, it was super helpful, this should be better now, lmk if you see anything else that can be improved